src/object_functions.c

```diff
@@ -305,45 +297,117 @@ void obj_loop_fireball_octoweapon(Object *obj, s32 speed) {
         }
     } else {
         if (obj->unk7C.word < 0) {
-            obj->unk7C.word += speed;
+            obj->unk7C.word += updateRate;
             if (obj->unk7C.word >= 0) {
                 obj->unk7C.word = 0;
             }
         } else {
-            obj->unk7C.word -= speed;
+            obj->unk7C.word -= updateRate;
             if (obj->unk7C.word <= 0) {
                 obj->unk7C.word = 0;
             }
         }
         if (obj->unk7C.word == 0) {
-            temp = obj64->unk1C;
+            temp = weapon->unk1C;
             if (temp != 0) {
                 func_8000488C(temp);
             }
-            func_80009558(SOUND_POP, obj->segment.trans.x_position, obj->segment.trans.y_position, obj->segment.trans.z_position, 4, 0);
+            func_80009558(SOUND_POP, obj->segment.trans.x_position, obj->segment.trans.y_position, obj->segment.trans.z_position, 4, NULL);
             gParticlePtrList_addObject(obj);
         }
     }
 }
-#else
-GLOBAL_ASM("asm/non_matchings/unknown_032760/obj_loop_fireball_octoweapon.s")
-#endif
 
+/**
+ * Laser gun init function.
+ * Sets the face direction and active behaviour modes based on entry data.
+ * This includes a targeting system and rate of fire.
+*/
 void obj_init_lasergun(Object *obj, LevelObjectEntry_Lasergun *entry) {
-    Object_LaserGun *obj64;
-    obj->unk4C->unk14 = 0x22;
+    Object_LaserGun *laserGun;
+    obj->unk4C->unk14 = 34;
     obj->unk4C->unk11 = 0;
-    obj64 = &obj->unk64->laser_gun;
-    obj64->unkE = entry->unkA;
-    obj64->unkF = entry->unkB;
-    obj64->unk10 = entry->unkC;
-    obj64->unk11 = entry->unkD;
-    obj64->unkC = obj64->unkF;
-    obj->segment.trans.y_rotation = entry->unk8 << 4 << 4; // Not sure about the shift amounts here, but it
-    obj->segment.trans.x_rotation = entry->unk9 << 4 << 4; // just needs to be 2 left shifts that add up to 8.
+    laserGun = &obj->unk64->lasergun;
+    laserGun->targeting = entry->targeting;
+    laserGun->fireRate = entry->fireRate;
+    laserGun->unk10 = entry->unkC;
+    laserGun->unk11 = entry->unkD;
+    laserGun->fireTimer = laserGun->fireRate;
+    obj->segment.trans.y_rotation = entry->angleX << 4 << 4; // Not sure about the shift amounts here, but it
+    obj->segment.trans.x_rotation = entry->angleY << 4 << 4; // just needs to be 2 left shifts that add up to 8.
 }
 
-GLOBAL_ASM("asm/non_matchings/unknown_032760/obj_loop_lasergun.s")
+/**
+ * Laser gun loop behaviour.
+ * A static mounted gun that shoots laserbeams at wherever it's facing.
+ * The fire rate is adjustable per object and can be set to target racers.
+*/
+void obj_loop_lasergun(Object* obj, s32 updateRate) {
```

Comment:
```suggestion
void obj_loop_lasergun(Object *obj, s32 updateRate) {
```
Someone needs to yell at me when I leave these on decomp.me